### DISCUSSION TITLE
Nice generator, but a couple questions

### DISCUSSION THREAD

**Discussion by CyberBotX** [canton7/PropertyChanged.SourceGenerator#8] [Q&A] [answered]
Hi there. I started using this source generator to replace using PostSharp's NotifyPropertyChanged support and I think it works out great. But I do have a few things I wanted to point out.

First, it seems that the source generator works fully when compiling but during design-time, it seems to break after the first time of applying it to a field. As in, the first field has the proper Intellisense and the generated source is available to see. But once I apply it on a second field, suddenly Visual Studio thinks that the generated code is gone, Intellisense no longer sees the attribute or the generated property. I don't know if this is a problem because of how this source generator works or some other reason.

Second, it seems that the source generator does not transfer the XML documentation from the field to the generated property. Was it intended that the documentation was not transferred or is this something that could be done?

I do like using the source generator as I think it is better than having PostSharp try to add all its INotifyPropertyChanged post-compile and also that their support has a few issues here and there.

**Comment by canton7** [comment]
> First, it seems that the source generator works fully when compiling but during design-time, it seems to break after the first time of applying it to a field. As in, the first field has the proper Intellisense and the generated source is available to see. But once I apply it on a second field, suddenly Visual Studio thinks that the generated code is gone, Intellisense no longer sees the attribute or the generated property

Interesting, I haven't seen anything like that myself.

When you say that VS thinks the generated code is gone, are you looking at the generated code under the 'Analysers' node of the Solution Explorer, or are you referring to something else?

If it works properly during a full build, I rather suspect that any design-time issues have to be a Visual Studio thing. The source generator itself doesn't know the difference between a full build and design-time -- it just gets called on a Compilation.

> Second, it seems that the source generator does not transfer the XML documentation from the field to the generated property. Was it intended that the documentation was not transferred or is this something that could be done?

I hadn't even considered that. Should be an easy fix, thanks.

**Reply by CyberBotX** [reply]
So when everything is working right, I see all the generated files when I expand the node under `Dependencies > Analyzers > PropertyChanged.SourceGenerator` in my project. But when I add a new field with the `[Notify]` attribute on it, the node then shows "This generator is not generating files." and that is when it acts as if the Notify attribute isn't defined and that none of the properties are defined. It is only broken during design-time, though, it compiles fine regardless of that. Everything works again if I restart Visual Studio. I'm using VS 2022 currently. I suppose it could be that the tooling for source generators still has some issues. But I don't have issues with other source generators.

**Reply by canton7** [reply]
Interesting, I'll have a shot at reproing, thanks for the details. It might be a week or two before I have time though

**Reply by CyberBotX** [reply]
That's fine. One thing I might suggest, although I haven't written my own SGs and only found this via searching, is maybe using `IIncrementalGenerator` instead of `ISourceGenerator`. I've no idea if this would change anything, though. I think it is also a larger change than just changing the interface.

**Reply by canton7** [reply]
Yeah I was looking into that. There's a bit of re-architecting required (there's already a nice spilt between the analysis and generation already which works well with incremental SGs). The main pain points are going to be 1) The EventArgs cache: the generated code for one class depends on the generated code for all other classes, which won't play well with incremental SGs, and 2) We need to analyse both the semantic model (for most of our info) and the syntax tree (to tell whether a type is partial), which complicates things a bit.

**Reply by canton7** [reply]
So, I managed to reproduce the "This generator is not generating files" thing. Interestingly, it doesn't generate Attribute.cs either, and that's entirely independent of all of the other generation

I found this issue, which looks similar: https://github.com/dotnet/aspnetcore/issues/38238. Indeed, adding `<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>` shows that the SG is generating updated files as you add new `[Notify]` properties, as they're present in obj/generated. The problem does seem to be that the VS solution explorer and Intellisense aren't picking them up for some reason. 

I'm not sure what I'm doing to trip over this issue -- I'm certain this wasn't happening the version of VS I used to initially develop and test the SG.

**Reply by canton7** [reply]
Since then, I've had a lot of problems with Microsoft-authored Source Generators hitting this same problem, and have seen other reports on StackOverflow/etc of other users hitting the same, so I'm pretty confident this isn't a problem with PropertyChanged.SourceGenerator.

It also doesn't seem like incremental SGs handle generating from class hierarchies, which would be a hard stop on this SG using them. Discussion here: https://github.com/dotnet/roslyn/discussions/60273

**Comment by canton7** [comment, marked as answer]
> Should be an easy fix, thanks.

https://github.com/canton7/PropertyChanged.SourceGenerator/commit/1d40b8ff5e03df331f30fc07d0bf3eea8e5efa6d

I'll hold off on a new release for now, in case anything needs to be done for your first point.

**Reply by canton7** [reply]
This has now been released.

**Reply by CyberBotX** [reply]
So I just updated the NuGet package in one of my projects to 1.0.2 and restarted Visual Studio, it gives me the "This generator is not generating files." message in the Solution Explorer and treats every place where I used the `[Notify]` attribute as an error, but it still compiles fine. I've done nothing since restarting Visual Studio for this project.

**Reply by canton7** [reply]
Restart VS again? Sounds like https://github.com/dotnet/roslyn/issues/49249, which has started biting me a lot recently.

**Reply by CyberBotX** [reply]
I have restarted it multiple times now, even shutting down all instances before starting up the project I updated. It still keeps giving me that message and shows no files but compiles fine.

**Reply by canton7** [reply]
Try adding `<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles> ` to your csproj, and look for generated files in obj?

My recollection is that the files keep being generated there, it's a VS issue that generated files aren't actually picked up by the editor.

Also try downgrading again and see what happens?

**Reply by CyberBotX** [reply]
Adding that to my csproj does show the generated files there. Downgrading the package back to 1.0.1 (and restarting Visual Studio, without the extra tag in my csproj) brings things back to working again, the files show up under the Solution Explorer, no errors at design-time.

Something I forgot to bring up, when I initially upgraded the package to 1.0.2, before I restarted Visual Studio, the arrow next to PropertyChanged.SourceGenerator under Analyzers disappeared. That's what made me restart Visual Studio the first time and when I started seeing the errors. I'm not sure if that is relevant considering how much that Roslyn issue seems to be causing.

**Reply by canton7** [reply]
If the files are being generated (as you can see on disk), then the SG is running correctly. Any problems with VS not using them I think have to be a VS issue.

I do wonder... There's a report on that Roslyn issue that a single crash in the SG will stop it working permanently. If PropertyChanged.SourceGenerator is crashing on some invalid input, if only for a microsecond before it is run again on valid input and starts working again, that might be enough to trigger the Roslyn issue. But the crash would probably never be visible: just the secondar effect it has on Roslyn.

I'm failing to reproduce this issue with PropertyChanged.SourceGenerator. If you can narrow down the input to something minimal which reproduces this, maybe I can use that as a starting point to investigate?

**Reply by CyberBotX** [reply]
So in the process of trying to make a minimal reproduction, I wasn't able to do so, but I did notice some times where while I was typing to create the minimal reproduction, the source generator would suddenly stop generating files (like my initial problem). But restarting Visual Studio brought everything back. I might not have a chance to do this until after my work day, but I might attempt to clone the GitHub repo and build the SG locally within my project with the problem and try to attach a debugger to see if something is going on.

**Reply by canton7** [reply]
Sounds good, thanks. My plan was to stick a catch block [here](https://github.com/canton7/PropertyChanged.SourceGenerator/blob/4639794683b4ebcb1ac042ca991810b28322196a/src/PropertyChanged.SourceGenerator/PropertyChangedSourceGenerator.cs#L58) with a `Debugger.Break()` inside. If that's hit, that supports my theory that it's a one-off exception that's tripping up Roslyn. If swallowing all exceptions fixes Roslyn, then I think my theory's correct.

Note that you need to restart VS whenever the SG is changed, because of https://github.com/dotnet/roslyn/issues/48083.

**Reply by canton7** [reply]
So it looks like my hunch was probably right: https://github.com/dotnet/roslyn/issues/49249#issuecomment-1079267139

I'd still like to understand what exception PropertyChanged.SourceGenerator is throwing here, as it's probably indicative of something I'm mis-handling. So if you have any luck reproing (and I think my suggestion of putting a `Debugger.Break()` inside a `catch` there is still a good one), or you manage to create a repro which I can run, I'm still very interested!

**Reply by CyberBotX** [reply]
So I added a `Debugger.Launch()` (so I could have it launch a new debugger instance as soon as something happened) and was getting an XML exception about root element not found (I didn't copy the exception at the time because I wanted to dig into it more first). So after digging in more (by launching the debugger before the try block and stepping through lines one at a time until it hit the exception), turns out it was one of my fields that had `[Notify]` on it had ill-formed XML documentation in the form of a missing `</summary>`. So technically it was my fault, not the generator's fault. Once I fixed that, everything started working as expected, but I can see why it wasn't failing with 1.0.1 but was with 1.0.2, since you didn't have it parsing XML documentation before.

Still a shame that the current version of Visual Studio can't cope with an exception being thrown by a source generator and also gives no errors in the Error List. Good to see that they'll have it fixed in the future, though.

**Reply by canton7** [reply]
Nice find, thank you!

I'll fix the xml parsing thing, and I'll also catch all exceptions rather than letting them bubble up to Roslyn (and raise a diagnostic)

**Reply by canton7** [reply]
Fixed, and I've tweaked the behaviour if we get an unexpected exception while processing a type: we now only fail to generate that type and its children, and we report the exception ourselves rather than chucking it up to Roslyn.

I'll hold off on the release for a couple of days: got one or two more things I want to squeeze into it.

**Reply by canton7** [reply]
Right, that's released.
